src/update_agent/mod.rs

```diff
@@ -195,9 +222,70 @@ impl UpdateAgentState {
         self.transition_to(target);
     }
 
-    /// Transition to the UpdateStaged state.
+    /// Transition to the UpdateStaged state, setting the number of postponements
+    /// to 0.
     fn update_staged(&mut self, update: Release) {
-        let target = UpdateAgentState::UpdateStaged(update);
+        let target = UpdateAgentState::UpdateStaged((update, 0));
+
+        self.transition_to(target);
+    }
+
+    /// Postpone an update finalization depending on whether interactive sessions
+    /// are present and whether the maximum number of allowed postponements has
+    /// been reached. Also possibly broadcast a warning message to logged in
+    /// interactive users.
+    ///
+    /// Returns a boolean indicating whether the a postponement has been registered
+    /// in the agent state.
+    fn handle_interactive_sessions(&mut self, interactive_sessions: &[InteractiveSession]) -> bool {
+        // If no users logged in interactively, don't postpone.
+        if interactive_sessions.is_empty() {
+            return false;
+        }
+
+        let (release, postponements) = match self.clone() {
+            UpdateAgentState::UpdateStaged((r, p)) => (r, p),
+            _ => unreachable!(
+                "transition not allowed: handle_interactive_sessions on {:?}",
+                self,
+            ),
+        };
+
+        if postponements >= MAX_FINALIZE_POSTPONEMENTS {
+            return false;
+        }
+
+        self.broadcast_reboot_warning(postponements, &release.version, interactive_sessions);
+        self.reboot_postponed(release, postponements.saturating_add(1));
+        UPDATE_FINALIZATION_POSTPONEMENTS.inc();
+
+        true
+    }
+
+    /// Broadcast a warning to interactive_sessions about an imminent reboot
+    /// upon first and last postponement.
+    fn broadcast_reboot_warning(
+        &mut self,
+        postponements: u8,
+        release_ver: &str,
+        interactive_sessions: &[InteractiveSession],
+    ) {
+        let warning_msg;
+        if postponements == 0 {
+            let max_reboot_delay_sec =
+                (MAX_FINALIZE_POSTPONEMENTS as u64).saturating_mul(DEFAULT_POSTPONEMENT_TIME_SECS);
```

Comment:
This feels a bit odd to read. Did you try inverting the direction of `postponements` counter, going from MAX to 0, so that it gives you directly the  correct number?

Reply:
Hmm yeah that would spare us from having to `MAX_FINALIZE_POSTPONEMENTS`.saturating_sub(1)` and instead just check of whether `postponements_left == 1`.